libs/gltfio/src/MaterialGenerator.cpp

```diff
@@ -0,0 +1,316 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "MaterialGenerator.h"
+
+#include <filamat/MaterialBuilder.h>
+
+#include <utils/Log.h>
+
+#include <string>
+
+using namespace filamat;
+using namespace filament;
+using namespace utils;
+
+namespace gltfio {
+namespace details {
+
+bool MaterialGenerator::EqualFn::operator()(const MaterialKey& k1, const MaterialKey& k2) const {
+    return
+        (k1.doubleSided == k2.doubleSided) &&
+        (k1.unlit == k2.unlit) &&
+        (k1.hasVertexColors == k2.hasVertexColors) &&
+        (k1.hasBaseColorTexture == k2.hasBaseColorTexture) &&
+        (k1.hasMetallicRoughnessTexture == k2.hasMetallicRoughnessTexture) &&
+        (k1.hasNormalTexture == k2.hasNormalTexture) &&
+        (k1.hasOcclusionTexture == k2.hasOcclusionTexture) &&
+        (k1.hasEmissiveTexture == k2.hasEmissiveTexture) &&
+        (k1.alphaMode == k2.alphaMode) &&
+        (k1.baseColorUV == k2.baseColorUV) &&
+        (k1.metallicRoughnessUV == k2.metallicRoughnessUV) &&
+        (k1.emissiveUV == k2.emissiveUV) &&
+        (k1.aoUV == k2.aoUV) &&
+        (k1.normalUV == k2.normalUV) &&
+        (k1.alphaMaskThreshold == k2.alphaMaskThreshold);
+}
+
+MaterialGenerator::MaterialGenerator(Engine* engine) : mEngine(engine) {
+    MaterialBuilder::init();
+}
+
+size_t MaterialGenerator::getMaterialsCount() const noexcept {
+    return mMaterials.size();
+}
+
+const Material* const* MaterialGenerator::getMaterials() const noexcept {
+    return mMaterials.data();
+}
+
+void MaterialGenerator::destroyMaterials() {
+    for (auto& iter : mCache) {
+        mEngine->destroy(iter.second);
+    }
+    mMaterials.clear();
+    mCache.clear();
+}
+
+static std::string shaderFromKey(const MaterialKey& config, const UvMap& uvmap) {
+    const auto normalUV = std::to_string(uvmap[config.normalUV] - 1);
+    const auto baseColorUV = std::to_string(uvmap[config.baseColorUV] - 1);
+    const auto metallicRoughnessUV = std::to_string(uvmap[config.metallicRoughnessUV] - 1);
+    const auto emissiveUV = std::to_string(uvmap[config.emissiveUV] - 1);
+    const auto aoUV = std::to_string(uvmap[config.aoUV] - 1);
+
+    std::string shader = R"SHADER(
+
+        // Sigh, assimp flips texture coords in its glTF2Importer, but we're not using assimp so we
+        // need to flip them here.
+        #if defined(HAS_ATTRIBUTE_UV0)
+        float2 uv0() { vec2 uv = getUV0(); uv.y = 1.0 - uv.y; return uv; }
```

Comment:
Could you use the `MaterialBuilder::flipUV` material property here instead?

Reply:
Whoa nice!